Add JWT-based authentication with backward compatibility

ztripez opened the pull request:
## Summary

This PR introduces JWT-based authentication tokens for Music Assistant while maintaining full backward compatibility with existing hash-based tokens. The implementation provides a foundation for future OIDC integration and claims-based permissions.

## Changes

### New Files
- `music_assistant/helpers/jwt_auth.py`: JWT encoding/decoding helper class
  - HS256 algorithm for token signing
  - Configurable secret key (auto-generated on first run)
  - Token encoding with standard JWT claims (sub, jti, iat, exp)
  - User metadata in claims (username, role, filters)
  - Token ID extraction without full validation

### Modified Files
- `music_assistant/controllers/webserver/auth.py`:
  - Updated `create_auth_token()` to generate JWT tokens
  - Updated `authenticate_with_token()` to accept both JWT and legacy tokens
  - Database expiration remains source of truth for both token types
  - Sliding window refresh for short-lived tokens (<30 days)

- `music_assistant/controllers/config.py`:
  - Added JWT secret key configuration entry
  - Auto-generates secure secret key on startup if not configured

## Token Format

JWT tokens include the following claims:
- `sub`: User ID
- `jti`: Token ID (for revocation via database)
- `iat`: Issued at timestamp
- `exp`: Expiration timestamp
- `username`: Username
- `role`: User role (admin/user)
- `player_filter`: Player filter configuration
- `provider_filter`: Provider filter configuration  
- `token_name`: Human-readable token name
- `is_long_lived`: Whether token is long-lived (>30 days)

## Backward Compatibility

- All existing hash-based tokens continue to work
- Token type auto-detected by format (JWT has 3 parts separated by dots)
- Database structure unchanged
- API endpoints unchanged
- Frontend requires no changes

## Security

- Tokens signed with HS256 algorithm using 256-bit secret key
- Secret key auto-generated using `secrets.token_urlsafe(32)`
- Database expiration checked as source of truth (prevents token reuse after revocation)
- Token ID (jti) allows individual token revocation
- No breaking changes to authentication flow

## Testing

All existing authentication tests pass (36/36):
- Hash token authentication
- JWT token authentication  
- Token creation and validation
- Token expiration handling
- Token revocation
- Sliding window refresh for short-lived tokens

## Future Work

This PR provides the foundation for:
- Claims-based permissions (follow-up PR)
- External OIDC provider support
- Provider-contributed custom claims
- OAuth2 refresh token flow

Commit:
Add JWT-based authentication with backward compatibility

Migrates from hash-based tokens to JWT (JSON Web Tokens) while maintaining
full backward compatibility with existing tokens. This enables stateless
authentication with embedded user claims for better integration with
external systems and OAuth2/OIDC compliance.

JWT Claims Structure:
- Standard claims: sub (user_id), jti (token_id), iat, exp
- Custom claims: username, role, player_filter, provider_filter,
  token_name, is_long_lived

Token Types:
- Short-lived (30 days, auto-renewing on use, sliding window)
- Long-lived (10 years, no auto-renewal for API integrations)

Implementation:
- JWTHelper class for encoding/decoding with HS256 algorithm
- JWT secret key generated and stored in auth.db settings table
- Token verification tries JWT first, falls back to legacy hash lookup
- Database still stores tokens for revocation checking

Benefits:
- Stateless authentication (user info embedded in token)
- Permission scopes available without database lookup
- OAuth2/OIDC compatibility for external integrations
- Standard JWT format for third-party verification

Migration Strategy:
- Automatic: Old tokens work until expiration
- New logins get JWT tokens automatically
- No breaking changes for existing clients

Commit:
Fix JWT token expiration check to honor database expiration

Database expiration is the source of truth for token validity, not just
the JWT expiration claim. This ensures manual token expiration (via
database update) works correctly even when JWT exp is still valid.

Commit:
Add documentation for future OIDC support

Notes on consuming external OIDC vs acting as OIDC provider, and
refresh token requirements for the latter.

Commit:
Clean up JWT implementation: remove dead code and verbose comments

- Remove unused methods: refresh_short_lived_token(), get_user_from_token()
- Remove unused imports: timedelta, UserRole
- Move User import to TYPE_CHECKING block
- Remove TODO comment about refresh tokens (not implementing)
- Simplify inline comments and reduce verbosity
- All tests still passing (36/36)

marcelveldt commented:
Please let me know what the fundamental issue is that this PR/feature tries to solve ?
I'd like to get a sales pitch why we really need to add this extra layer of complexity that needs to be maintained.

Don't get me wrong, I'm delighted that people are willing to invest time by doing PR's to extend functionality to core functionality but I kept the auth system simple on purpose to keep the scope limited. With adding more functionality, we increase maintenance burden and possible security issues.

ztripez commented:
> Please let me know what the fundamental issue is that this PR/feature tries to solve ? I'd like to get a sales pitch why we really need to add this extra layer of complexity that needs to be maintained.
> 
> Don't get me wrong, I'm delighted that people are willing to invest time by doing PR's to extend functionality to core functionality but I kept the auth system simple on purpose to keep the scope limited. With adding more functionality, we increase maintenance burden and possible security issues.

The main problem this addresses is that the current auth model is very coarse. As soon as a provider needs anything more granular than admin vs user, it has to invent its own permission checks. That is what happened with my MCP provider.

This PR does not change the end-user model and does not introduce per-user permissions or RBAC. Admin and user still exist as roles. What changes is that roles are expressed as predefined claim sets, and functionality can declare what it requires.

I want to be clear that I am not taking this the wrong way. I agree with you that the permission system needs to be kept on a very tight leash, because it can snowball quickly and become hard to maintain. That is also why I am explicitly not proposing user-level permission management inside MA.

The claims approach keeps MA simple while giving providers a consistent way to do fine-grained authorization, instead of each provider solving this differently. If someone wants real per-user permission granularity, my view is that this should be handled by an external identity provider via OIDC rather than modeled inside MA itself.

From an end-user perspective, nothing changes. From a provider and platform perspective, this avoids ad-hoc permission systems and makes authorization more explicit and predictable.

marcelveldt commented on music_assistant/helpers/jwt_auth.py:
```diff
@@ -0,0 +1,111 @@
+            "username": user.username,
+            "role": user.role.value,
+            "player_filter": user.player_filter,
+            "provider_filter": user.provider_filter,
```
What is the advantage of having this stored in the token ?
I mean, it can dynamically be updated and then the token still has the old info.

ztripez replied:
Hm.. yes that is correct, and any future more fine grained access can be done with claims. I'll remove it.

MarvinSchenkel replied:
@ztripez Could you have a look at this so we can get this merged please?

marcelveldt commented:
We should consider switching to JWT by default so we can phase out the legacy tokens at some point.
So, in a follow-up PR, adjust the code to hand out JWT tokens by default and update the python client lib and frontend code.
Then we only accept legacy tokens (for a while, like 1 year or so) for token-auth only but login will always use JWT.
Does that make sense ?

Commit:
Remove player_filter and provider_filter from JWT claims

These values can be dynamically updated, so storing them in the token
would result in stale data. The current values are available from the
database lookup during token validation.

MarvinSchenkel reviewed (approved):
Thanks @ztripez 🎉
